src/org/zaproxy/zap/authentication/PostBasedAuthenticationMethodType.java

```diff
@@ -320,47 +326,63 @@ public WebSession authenticate(SessionManagementMethod sessionManagementMethod,
 			}
-			List<AntiCsrfToken> antiCsrfTokensOfFirstLoginMsg = null;
+			List<AntiCsrfToken> freshAcsrfTokens = null;
   			if(extAntiCsrf != null) {
-				antiCsrfTokensOfFirstLoginMsg = extAntiCsrf.getTokens(requestMessage);
-			}
-			AntiCsrfToken antiCsrfTokenOfFirstLoginMsg = null;
-			if(antiCsrfTokensOfFirstLoginMsg != null && antiCsrfTokensOfFirstLoginMsg.size() > 0) {
-				antiCsrfTokenOfFirstLoginMsg = antiCsrfTokensOfFirstLoginMsg.get(0);
+				freshAcsrfTokens = extAntiCsrf.getTokensFromResponse(loginMsgWithFreshAcsrfToken);
+			} else {
+				LOGGER.warn("ExtensionAntiCSRF is not available, skipping ACSRF replacing task");
+				return;
 			}
-			
-			if (antiCsrfTokenOfFirstLoginMsg != null) {
-				HttpMessage loginMsgToRegenerateAntiCsrfToken = antiCsrfTokenOfFirstLoginMsg.getMsg().cloneAll();
-				loginMsgToRegenerateAntiCsrfToken.setRequestingUser(user);
-				
-				// Clear any session identifiers
-				loginMsgToRegenerateAntiCsrfToken.getRequestHeader().setHeader(HttpRequestHeader.COOKIE, null);
+			if(freshAcsrfTokens != null && freshAcsrfTokens.size() > 0) {
+				if(LOGGER.isDebugEnabled()) {
+					LOGGER.debug("The login page has " + freshAcsrfTokens.size() + " ACSRF token(s)");
+				}
 				
-				getHttpSender().sendAndReceive(loginMsgToRegenerateAntiCsrfToken);
-				AuthenticationHelper.addAuthMessageToHistory(loginMsgToRegenerateAntiCsrfToken);
-
-				String regeneratedAntiCsrfTokenValue = extAntiCsrf.getTokenValue(loginMsgToRegenerateAntiCsrfToken, antiCsrfTokenOfFirstLoginMsg.getName());
-				String requestBody = requestMessage.getRequestBody().toString();
-				if (regeneratedAntiCsrfTokenValue != null) {
-					requestBody = requestBody.replace(
-							antiCsrfTokenOfFirstLoginMsg.getValue(), paramEncoder.apply(regeneratedAntiCsrfTokenValue));
-					requestMessage.getRequestBody().setBody(requestBody);
+				String postRequestBody = requestMessage.getRequestBody().toString();
+				Map<String, String> parameters = extractParametersFromPostData(postRequestBody);
+				if(parameters != null) {
+					String oldAcsrfTokenValue = null;
+					String replacedPostData = null;
+					for (AntiCsrfToken antiCsrfToken : freshAcsrfTokens) {
+						oldAcsrfTokenValue = parameters.get(antiCsrfToken.getName());
+						replacedPostData = postRequestBody.replace(oldAcsrfTokenValue, antiCsrfToken.getValue());
+						
+						if(LOGGER.isDebugEnabled()) {
+							LOGGER.debug("replaced " + oldAcsrfTokenValue + " old ACSRF token value with " + antiCsrfToken.getValue());
+						}
+					}
+					requestMessage.getRequestBody().setBody(replacedPostData);
+				} else {
+					LOGGER.warn("ACSRF token found but could not replace old value with fresh value");
 				}
+			} else {
+				if(LOGGER.isDebugEnabled()) {
+					LOGGER.debug("No ACSRF token found in the response of " + loginMsgWithFreshAcsrfToken.getRequestHeader());
+				}
+			}
+		}
+		
+		private Map<String, String> extractParametersFromPostData(String postRequestBody) {
```

Comment:
Ideally this should be delegated to the implementations as they handle different formats (using `extractParameters` and `replaceParameterValue`? although these would have to be moved around, from options to the auth method), fine if handled in a follow up PR though.